lib/kernel/src/code_server.erl

```diff
@@ -1409,7 +1405,7 @@ handle_on_load(Res, Action, _, _, St) ->
 handle_pending_on_load(Action, Mod, From, #state{on_load=OnLoad0}=St) ->
     case lists:keyfind(Mod, 2, OnLoad0) of
 	false ->
-	    Action(ok, St);
+	    Action({module, Mod}, St);
 	{{From,_Ref},Mod,_Pids} ->
 	    %% The on_load function tried to make an external
 	    %% call to its own module. That would be a deadlock.
```

Comment:
I think in this case it still won't work correctly - this bypasses the call to `reply_loading` and just returns `reply` tuple directly.
Should this branch be changed to call `Action({error, deadlock}, St)` instead?

Reply:
I could not run into this branch in my tests. It is either called if there is no waiting list (and then a single reply is fine) or skipped altogether if there is a waiting list. It also felt inaccurate to return `{error, deadlock}`. It probably makes sense only for the caller itself and the others should still see it as nofile/badfile?

Reply:
My understanding is that it could trigger for an `on_load` function that is slow & recursive. I can generate a hanging situation with code like:+
```
-module(slow_on_load).

-on_load(on_load/0).

on_load() ->
    timer:sleep(1000),
    slow_on_load:foo().

foo() -> ok.
```

and calling `spawn(fun() -> code:ensure_loaded(slow_on_load) end), code:ensure_loaded(slow_on_load).` from the shell.
This is with code from this PR checked out.

Reply:
Unfortunately, the error is not happening on this line of code. In your code, we never get to this clause, because we are stuck on loading the module.

I have a fix for this, but it is a bigger change, and I am not sure how comfortable I am in adding it to maint. My suggestion is to merge this into maint.

To recap:

1. If you have a deadlock inside an on_load callback by calling yourself, it will no longer go detected (it will effectively deadlock)
2. While this is a regression, previously deadlocking code would most likely fail anyway, so this regression is unlikely to break any actual code.
3. We should merge this PR into maint.
4. I will send another PR later which the team can evaluate if they want to bring into maint or not.

Reply:
Of course, as soon as I wrote the above, I found a clean solution to the problem. I just had to move the `handle_pending_on_load` call up. I pushed a new commit for easier review, I'd be glad to squash or submit an additional PR. :)